realm/src/main/java/io/realm/RealmList.java

```diff
@@ -183,6 +191,10 @@ public E set(int location, E object) {
 
     // Transparently copies a standalone object or managed object from another Realm to the Realm backing this RealmList.
     private E copyToRealmIfNeeded(E object) {
+        if (!(realm instanceof Realm))  {
+            throw new IllegalStateException("Not supported for RealmList containing DynamicRealmObjects");
+        }
+        Realm realm = (Realm) this.realm;
```

Comment:
shouldn't you update the object Realm? 
`object.realm=this.realm` 
+ comment/name is confusing this is not a copy the implementation reads more like a `moveTo`